channeldb/migration12/invoices.go

```diff
@@ -339,7 +361,73 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
 //
 // nolint: dupl
 func serializeInvoice(w io.Writer, i *Invoice) error {
-	panic("unimplemented")
+	creationDateBytes, err := i.CreationDate.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	settleDateBytes, err := i.SettleDate.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	var fb bytes.Buffer
+	err = i.Terms.Features.EncodeBase256(&fb)
+	if err != nil {
+		return err
+	}
+	featureBytes := fb.Bytes()
+
+	preimage := [32]byte(i.Terms.PaymentPreimage)
+	value := uint64(i.Terms.Value)
+	cltvDelta := uint32(i.FinalCltvDelta)
+	expiry := uint64(i.Expiry)
+
+	amtPaid := uint64(i.AmtPaid)
+	state := uint8(i.Terms.State)
+
+	tlvStream, err := tlv.NewStream(
+		// Memo and payreq.
+		tlv.MakePrimitiveRecord(memoType, &i.Memo),
+		tlv.MakePrimitiveRecord(payReqType, &i.PaymentRequest),
+
+		// Add/settle metadata.
+		tlv.MakePrimitiveRecord(createTimeType, &creationDateBytes),
+		tlv.MakePrimitiveRecord(settleTimeType, &settleDateBytes),
+		tlv.MakePrimitiveRecord(addIndexType, &i.AddIndex),
+		tlv.MakePrimitiveRecord(settleIndexType, &i.SettleIndex),
+
+		// Terms.
+		tlv.MakePrimitiveRecord(preimageType, &preimage),
```

Comment:
I start to realize that this migration offers more opportunities to clean up some things. The situation with "preimage unknown" could be dealt with much more elegant by just not writing out the preimage tlv record.

For the add/settle metadata similar a similar thing could be possible.

I know we want to move on with the rest of the stack of mpp prs, but the effort required to fix these things now seems to be much less than when it would need to be a separate migration later on. What do you think?

Reply:
i would vote to revisit these things if they become an issue, since those optimizations don't provide any tangible benefits in the grand scheme of things